ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/BlockAttestationsPool.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.statetransition;
+
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.getAttesterIndicesIntoCommittee;
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.isSingleAttester;
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.representsNewAttesterAggregate;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.PriorityBlockingQueue;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.storage.ChainStorage;
+import tech.pegasys.artemis.util.SSZTypes.Bitlist;
+import tech.pegasys.artemis.util.SSZTypes.SSZList;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class BlockAttestationsPool {
+
+  private final ConcurrentHashMap<Bytes32, UnsignedLong> dataRootToSlot = new ConcurrentHashMap<>();
+
+  private final ConcurrentHashMap<Bytes32, Bitlist> unprocessedAttestationsBitlist =
+      new ConcurrentHashMap<>();
+
+  private final ConcurrentHashMap<Bytes32, Bitlist> processedAttestationsBitlist =
+      new ConcurrentHashMap<>();
+
+  private final int QUEUE_MAX_SIZE =
+      Constants.MAX_VALIDATORS_PER_COMMITTEE
+          * Constants.MAX_COMMITTEES_PER_SLOT
+          * Constants.SLOTS_PER_EPOCH;
+
+  private final Queue<Attestation> aggregateAttesationsQueue =
+      new PriorityBlockingQueue<>(QUEUE_MAX_SIZE, Comparator.comparing(a -> a.getData().getSlot()));
+
+  public void addUnprocessedAttestationToQueue(Attestation newAttestation) {
+    if (!isSingleAttester(newAttestation)) {
+
+      Bytes32 attestationDataHash = newAttestation.getData().hash_tree_root();
+      ChainStorage.get(attestationDataHash, this.unprocessedAttestationsBitlist)
+          .ifPresentOrElse(
```

Comment:
When checking / setting values on a concurrent map, it's best to use the utilities provided by the map to make atomic updates.  By splitting up the check for a pre-existing value and the setting of the value, we're opening up the possibility of a race condition.  We should look into reworking this logic to something like:
```
final Bitlist bitlist = unprocessedAttestationsBitlist
        .computeIfAbsent(attestationDataHash, (key) -> newAttestation.getAggregation_bits().copy());
final boolean updated = updateBitlist(bitlist, newAttestation.getAggregation_bits());
if (updated) {...}
```